lib/srv/db/cassandra/protocol/conn.go

```diff
@@ -0,0 +1,213 @@
+	"github.com/gravitational/trace"
+)
+
+// NewConn is used to create a new connection.
+func NewConn(rawConn net.Conn) *Conn {
+	return &Conn{
+		Conn:        rawConn,
+		frameCodec:  frame.NewRawCodec(),
+		segmentCode: segment.NewCodec(),
+	}
+}
+
+// Conn represent incoming client connection or ongoing connection to cassandra server.
+// Reading and Writing Frames/Packages needs to be done sequential because Conn implementation
+// is not thread safe. Conn package is used to intercept and preform custom Cassandra handshake
+// and in case of connection incoming connection to provide ability to audit incoming client packages.
+type Conn struct {
+	net.Conn
+	segmentCode       segment.Codec
+	frameCodec        frame.RawCodec
+	modernLayoutRead  bool
+	modernLayoutWrite bool
+}
+
+// ReadPacket is used to read packet from the connection.
+func (c *Conn) ReadPacket() (*Packet, error) {
+	var buff bytes.Buffer
+	tr := io.TeeReader(c.Conn, &buff)
+
+	fr, err := c.readWithModernLayout(tr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	c.maybeSwitchToModernLayout(fr)
+
+	return &Packet{
+		raw:   buff,
+		frame: fr,
+	}, nil
+}
+
+// WriteFrame is used to write frame to the connection.
+func (c *Conn) WriteFrame(outgoing *frame.Frame) error {
+	if err := c.writeFrameWithModernLayout(outgoing); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if startup, ok := outgoing.Body.Message.(*message.Startup); ok {
+		compression := startup.GetCompression()
+		c.frameCodec = frame.NewRawCodecWithCompression(client.NewBodyCompressor(compression))
+		c.segmentCode = segment.NewCodecWithCompression(client.NewPayloadCompressor(compression))
+	}
+	c.maybeSwitchToModernLayout(outgoing)
+	return nil
+}
+
+// readWithModernLayout is used to read frame from the connection.
+// If the connection is using modern framing layout, it will read segments.
+// Otherwise, it will read frames.
+func (c *Conn) readWithModernLayout(r io.Reader) (*frame.Frame, error) {
+	if c.modernLayoutRead {
+		v, err := c.readSegment(r)
+		return v, trace.Wrap(err)
+	}
+	fr, err := c.readFrame(r)
+	return fr, trace.Wrap(err)
+}
+
+// readFrame is used to read frame from the connection.
+// If read frame is a Startup frame, it will switch to modern framing layout and
+// update codec to use modern framing layout.
+func (c *Conn) readFrame(r io.Reader) (*frame.Frame, error) {
+	fr, err := c.frameCodec.DecodeFrame(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if startup, ok := fr.Body.Message.(*message.Startup); ok {
+		compression := startup.GetCompression()
+		c.frameCodec = frame.NewRawCodecWithCompression(client.NewBodyCompressor(compression))
+		c.segmentCode = segment.NewCodecWithCompression(client.NewPayloadCompressor(compression))
+		// If moderate framing layout is supported all received from a client after Startup message should
+		// use segment encoding.
+		c.modernLayoutRead = fr.Header.Version.SupportsModernFramingLayout()
```

Comment:
I am confused by the two booleans. Why we only set read here but not write? Should this also be included in `maybeSwitchToModernLayout` func ?

Reply:
The conn is used in both sides of connection `client->proxy` and `proxy->dbserver` where the switch to modern layout is triggered by two different cases.